### DISCUSSION TITLE
One to many relation using data-classes

### DISCUSSION THREAD

**Discussion by yannick1974** [sqlalchemy/sqlalchemy#9302] [Usage Questions]
Hi,

I'd like to have a one-to-many relation that is “nullable on both side”. Using the  [Nullable Many-to-One](https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#nullable-many-to-one) example from the documentation works :

```python
from __future__ import annotations
from sqlalchemy.orm import DeclarativeBase, Mapped,, Session
from sqlalchemy import create_engine, ForeignKey
from typing import List

from sqlalchemy.orm import mapped_column, relationship


class Base(DeclarativeBase):
    pass


class Parent(Base):
    __tablename__ = "parent_table"

    id: Mapped[int] = mapped_column(primary_key=True)
    child_id: Mapped[int | None] = mapped_column(
        ForeignKey("child_table.id")
    )
    child: Mapped[Child | None] = relationship(back_populates="parents")


class Child(Base):
    __tablename__ = "child_table"

    id: Mapped[int] = mapped_column(primary_key=True)
    parents: Mapped[List[Parent]] = relationship(back_populates="child")


engine = create_engine("sqlite://")

Base.metadata.create_all(engine)

with Session(engine) as session:
    p1 = Parent()
    c1 = Child()
    p2 = Parent(child=c1)

    session.add_all([c1, p1, p2])
    session.commit()
```

I tried to convert this to a _Declarative Dataclass Mapping_ like this:

```python
from __future__ import annotations
from sqlalchemy.orm import DeclarativeBase, Mapped, MappedAsDataclass, Session
from sqlalchemy import create_engine, ForeignKey
from typing import List

from sqlalchemy.orm import mapped_column, relationship


class Base(MappedAsDataclass, DeclarativeBase):
    pass


class Parent(Base):
    __tablename__ = "parent_table"

    id: Mapped[int] = mapped_column(primary_key=True, init=False)
    child_id: Mapped[int | None] = mapped_column(
        ForeignKey("child_table.id"), init=False
    )
    child: Mapped[Child | None] = relationship(back_populates="parents", default=None)


class Child(Base):
    __tablename__ = "child_table"

    id: Mapped[int] = mapped_column(primary_key=True, init=False)
    parents: Mapped[List[Parent]] = relationship(back_populates="child", default=None)


engine = create_engine("sqlite://")

Base.metadata.create_all(engine)

with Session(engine) as session:
    p1 = Parent()
    c1 = Child()
    p2 = Parent(child=c1)

    session.add_all([c1, p1, p2])
    session.commit()
```

But that fails with `TypeError: Incompatible collection type: None is not list-like` for `c1 = Child()`. 

How can I express that the parent attribute of the `Child` class may be null?

I tried
```python
parents: Mapped[List[Parent] | None] = relationship(back_populates="child", default=None) 
```
or using `Optional` but that raised the same error.

Of course, I could set the `parent` relation to `init=False` in the `Child` definition and only make the association while creating the parent but I'd like to be able to do both.

Do you have any idea?

Best,

Yannick

**Comment by zzzeek** [comment]
> class Parent(Base):
>     __tablename__ = "parent_table"
> 
>     id: Mapped[int] = mapped_column(primary_key=True, init=False)
>     child_id: Mapped[int | None] = mapped_column(
>         ForeignKey("child_table.id"), init=False
>     )
>     child: Mapped[Child | None] = relationship(back_populates="parents", default=None)
> 
> 
> class Child(Base):
>     __tablename__ = "child_table"
> 
>     id: Mapped[int] = mapped_column(primary_key=True, init=False)
>     parents: Mapped[List[Parent]] = relationship(back_populates="child", default=None)

this pattern isn't supported.  in SQL, one-to-many collections can be empty, but they can't be "NULL", there's nothing to be set to NULL.

**Comment by yannick1974** [comment]
Is doing like this the correct way to go? (at least it works).

```python
class Child(Base):
    __tablename__ = "child_table"

    id: Mapped[int] = mapped_column(primary_key=True, init=False)
    parents: Mapped[List[Parent]] = relationship(
        back_populates="child", default_factory=list
    )
 ```

**Reply by zzzeek** [reply]
seems right
